lib/SPIRV/SPIRVToOCL20.cpp

```diff
@@ -135,11 +135,33 @@ void SPIRVToOCL20::visitCallSPIRVControlBarrier(CallInst *CI) {
           return cast<ConstantInt>(Args[I])->getZExtValue();
         };
         auto ExecScope = static_cast<Scope>(GetArg(0));
-        auto MScope = static_cast<Scope>(GetArg(1));
-        auto Sema = mapSPIRVMemSemanticToOCL(GetArg(2));
+        auto MemScope = static_cast<Scope>(GetArg(1));
+
         Args.resize(2);
-        Args[0] = getInt32(M, Sema.first);
-        Args[1] = getInt32(M, rmap<OCLScopeKind>(MScope));
+        if (auto Arg = dyn_cast<ConstantInt>(Args[2])) {
+          auto Sema = mapSPIRVMemSemanticToOCL(Arg->getZExtValue());
+          Args[0] = getInt32(M, Sema.first);
+        } else {
+          CallInst *TransCall = dyn_cast<CallInst>(Args[2]);
+          Function *F = TransCall ? TransCall->getCalledFunction() : nullptr;
+          // In case the SPIR-V module was created from an OpenCL program by
+          // *this* SPIR-V generator, we know that the value passed to
+          // __translate_ocl_memory_fence is what we should pass to the OpenCL
+          // builtin now.
+          if (F && F->getName().equals(kSPIRVName::TranslateOCLMemFence)) {
```

Comment:
This is not quite true actually: we allow translating of non-constant memory semantic from SPIR-V to LLVM IR, but not vice-versa.

It is probably better to prepare a separate patch to address that, because there are several places similar to that one

Reply:
Ok, I though that there is at least `TranslateOCLMemFence` function and apparently it doesn't exists. I don't see a lot value then in wrong condition here: remove this kind of optimization completely